components/opponent/wikis/warcraft/opponent_custom.lua

```diff
@@ -0,0 +1,112 @@
+local String = require('Module:StringUtils')
+local Table = require('Module:Table')
+local TeamTemplate = require('Module:TeamTemplate')
+local TypeUtil = require('Module:TypeUtil')
+
+local Opponent = Lua.import('Module:Opponent', {requireDevIfEnabled = true})
+local PlayerExt = Lua.import('Module:Player/Ext/Custom', {requireDevIfEnabled = true})
+
+local CustomOpponent = Table.deepCopy(Opponent)
+
+CustomOpponent.types.Player = TypeUtil.extendStruct(Opponent.types.Player, {
+	race = 'string?',
+})
+
+CustomOpponent.types.PartyOpponent = TypeUtil.struct{
+	players = TypeUtil.array(CustomOpponent.types.Player),
+	type = TypeUtil.literalUnion(unpack(Opponent.partyTypes)),
+}
+
+CustomOpponent.types.Opponent = TypeUtil.union(
+	Opponent.types.TeamOpponent,
+	CustomOpponent.types.PartyOpponent,
+	Opponent.types.LiteralOpponent
+)
+
+function CustomOpponent.readOpponentArgs(args)
+	local opponent = Opponent.readOpponentArgs(args)
+	local partySize = Opponent.partySize((opponent or {}).type)
+
+	if partySize == 1 then
+		opponent.players[1].race = Faction.read(args.race)
+	elseif partySize then
+		for playerIx, player in ipairs(opponent.players) do
+			player.race = Faction.read(args['p' .. playerIx .. 'race'])
+		end
+	end
+
+	return opponent
+end
+
+function CustomOpponent.fromMatch2Record(record)
+	local opponent = Opponent.fromMatch2Record(record)
+
+	if Opponent.typeIsParty(opponent.type) then
+		for playerIx, player in ipairs(opponent.players) do
+			local playerRecord = record.match2players[playerIx]
+			player.race = Faction.read(playerRecord.extradata.faction) or Faction.defaultFaction
+		end
+	end
+
+	return opponent
+end
+
+function CustomOpponent.toLpdbStruct(opponent)
+	local storageStruct = Opponent.toLpdbStruct(opponent)
+
+	if Opponent.typeIsParty(opponent.type) then
+		for playerIndex, player in pairs(opponent.players) do
+			storageStruct.opponentplayers['p' .. playerIndex .. 'faction'] = player.race
+		end
+	end
+
+	return storageStruct
+end
+
+function CustomOpponent.fromLpdbStruct(storageStruct)
+	local opponent = Opponent.fromLpdbStruct(storageStruct)
+
+	if Opponent.partySize(storageStruct.opponenttype) then
+		for playerIndex, player in pairs(opponent.players) do
+			player.race = storageStruct.opponentplayers['p' .. playerIndex .. 'faction']
+		end
+	end
+
+	return opponent
+end
+
+function CustomOpponent.resolve(opponent, date, options)
+	options = options or {}
+	if opponent.type == Opponent.team then
+		return Opponent.resolve(opponent, date, options)
+	elseif Opponent.typeIsParty(opponent.type) then
+		for _, player in ipairs(opponent.players) do
+			if options.syncPlayer then
+				local hasRace = String.isNotEmpty(player.race)
+				PlayerExt.syncPlayer(player, {savePageVar = not Opponent.playerIsTbd(player)})
+				if not player.team then
+					player.team = PlayerExt.syncTeam(player.pageName, nil, {date = date})
+				end
+				player.race = (hasRace or player.race ~= Faction.defaultFaction) and player.race or nil
```

Comment:
Should the syncPlayer handle this (the race stuff)

Reply:
imo no since PlayerExt.syncPlayer is also used on places where team is not wanted/needed